orttraining/orttraining/core/session/training_session.cc

```diff
@@ -916,11 +1040,211 @@ common::Status TrainingSession::Run(const RunOptions& run_options, IOBinding& io
   return InferenceSession::Run(run_options, io_binding);
 }
 
+void TrainingSession::CreateBatchVariables(
+    IOBinding& io_binding,
+    IOBinding& sub_io_binding,
+    const size_t slice_id,
+    const size_t num_slices) {
+  auto& inputs = io_binding.GetInputs();
+  const auto& input_names = io_binding.GetInputNames();
+
+  ORT_ENFORCE(inputs.size() == input_names.size(), "\"input\" and their names are parallel. One input should have one name.");
+
+  auto has_element = [&](const std::vector<std::string> vector, const std::string element) {
+    auto it = std::find(vector.begin(), vector.end(), element);
+    if (it != vector.end()) {
+      return true;
+    } else {
+      return false;
+    }
+  };
+
+  // Slice input tensors.
+  for (size_t i = 0; i < inputs.size(); ++i) {
+    const auto name = input_names[i];
+    ORT_ENFORCE(pipeline_context_.sliced_axes[name] >= 0,
+                "Sliced axis of input \"", name, "\" must be non-negative but got ", pipeline_context_.sliced_axes[name]);
+    const size_t slice_axis = static_cast<size_t>(pipeline_context_.sliced_axes[name]);
+
+    if (has_element(pipeline_context_.sliced_tensor_names, name)) {
+      OrtValue sliced_value = SliceTensor(inputs[i], slice_id, slice_axis, num_slices, *this);
+      sub_io_binding.BindInput(name, sliced_value);
+    } else {
+      sub_io_binding.BindInput(name, inputs[i]);
+    }
+  }
+
+  auto& outputs = io_binding.GetOutputs();
+  const auto& output_names = io_binding.GetOutputNames();
+
+  ORT_ENFORCE(outputs.size() == output_names.size(), "\"output\" and their names are parallel. One output should have one name.");
+
+  // Slice output tensors.
```

Comment:
Can we pull out the common parts for the input and output slicing

Reply:
Sure.

---
In reply to: [527652439](https://github.com/microsoft/onnxruntime/pull/5815#discussion_r527652439) [](ancestors = 527652439)